sdk/src/main/java/io/opentelemetry/sdk/tags/TagMapSdk.java

```diff
@@ -0,0 +1,167 @@
+ */
+
+package io.opentelemetry.sdk.tags;
+
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.Tag;
+import io.opentelemetry.tags.TagKey;
+import io.opentelemetry.tags.TagMap;
+import io.opentelemetry.tags.TagMetadata;
+import io.opentelemetry.tags.TagValue;
+import io.opentelemetry.tags.unsafe.ContextUtils;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+class TagMapSdk implements TagMap {
+
+  // The types of the TagKey and Tag must match for each entry.
+  private final Map<TagKey, Tag> tags;
+  @Nullable private final TagMap parent;
+
+  /**
+   * Creates a new {@link TagMapSdk} with the given tags.
+   *
+   * @param tags the initial tags for this {@code TagMapSdk}.
+   * @param parent providing a default set of tags
+   */
+  private TagMapSdk(Map<? extends TagKey, ? extends Tag> tags, TagMap parent) {
+    this.tags = Collections.unmodifiableMap(new HashMap<>(checkNotNull(tags, "tags")));
+    this.parent = parent;
+  }
+
+  @Override
+  public Iterator<Tag> getIterator() {
+    Map<TagKey, Tag> combined = new HashMap<>(tags);
+    if (parent != null) {
+      for (Iterator<Tag> it = parent.getIterator(); it.hasNext(); ) {
+        Tag tag = it.next();
+        if (!combined.containsKey(tag.getKey())) {
+          combined.put(tag.getKey(), tag);
+        }
+      }
+    }
+    // Clean out any null values that may have been added by Builder.remove.
+    for (Iterator<Tag> it = combined.values().iterator(); it.hasNext(); ) {
+      if (it.next() == null) {
+        it.remove();
+      }
+    }
+
+    return Collections.unmodifiableCollection(combined.values()).iterator();
+  }
+
+  @Nullable
+  @Override
+  public TagValue getTagValue(TagKey tagKey) {
+    Tag tag = tags.get(tagKey);
+    if (tag != null) {
+      return tag.getValue();
+    } else {
+      return parent == null ? null : parent.getTagValue(tagKey);
+    }
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || !(o instanceof TagMapSdk)) {
+      return false;
+    }
+
+    TagMapSdk tagMapSdk = (TagMapSdk) o;
+
+    if (!tags.equals(tagMapSdk.tags)) {
+      return false;
+    }
+    return parent != null ? parent.equals(tagMapSdk.parent) : tagMapSdk.parent == null;
+  }
+
+  @Override
+  public int hashCode() {
+    int result = tags.hashCode();
+    result = 31 * result + (parent != null ? parent.hashCode() : 0);
+    return result;
+  }
+
+  static class Builder implements TagMap.Builder {
+    @Nullable private TagMap parent;
+    private final Map<TagKey, Tag> tags;
+
+    /**
+     * Create a new {@link TagMap.Builder} with the provided {@link TagMap} as the parent.
+     *
+     * @param parent TagMap used as base map.
+     */
+    Builder(TagMap parent) {
```

Comment:
Is this needed?

Reply:
Now that it's package accessible only, I don't think it's useful.  I removed it.